### PR TITLE
regenerate with latest pango gir file

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Depends on https://github.com/gtk-rs/gir-files/pull/90

### CODE REVIEW

File: pango/src/auto/attr_list.rs [hunk 1]
```diff
@@ -84,10 +95,26 @@ impl AttrList {
             ffi::pango_attr_list_update(self.to_glib_none().0, pos, remove, add);
         }
     }
+
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+    #[doc(alias = "pango_attr_list_from_string")]
+    pub fn from_string(text: &str) -> Option<AttrList> {
```
Comment (bilelmoussaoui) [inline pango/src/auto/attr_list.rs:102]: This should probably return a Result & have a FromStr implementation

File: pango/src/auto/enums.rs [hunk 2]
```diff
@@ -1004,6 +1010,8 @@ pub enum FontScale {
     Superscript,
     #[doc(alias = "PANGO_FONT_SCALE_SUBSCRIPT")]
     Subscript,
+    #[doc(alias = "PANGO_FONT_SCALE_SMALL_CAPS")]
+    SmallCaps,
```
Comment (bilelmoussaoui) [inline pango/src/auto/enums.rs:1014]: This addition is fine because the enum itself was added in 1.50

File: pango/src/auto/tab_array.rs [hunk 3]
```diff
@@ -92,4 +134,37 @@ impl TabArray {
...
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+    #[doc(alias = "pango_tab_array_sort")]
+    pub fn sort(&mut self) {
+        unsafe {
+            ffi::pango_tab_array_sort(self.to_glib_none_mut().0);
+        }
+    }
+
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
```
Comment (bilelmoussaoui) [inline pango/src/auto/tab_array.rs:147]: those functions probably shouldn't take mut self, but all the functions in this type seems to be doing so already so... I will leave cleaning that up to someone that needs it/knows how it should be used

File: pango/src/auto/enums.rs (hunk 2 above)
Comment (sdroege) [reply pango/src/auto/enums.rs:1014]: Adding variants to enums is also generally not a problem

File: pango/Gir.toml [hunk 4]
```diff
@@ -310,3 +313,36 @@ boxed_inline = true
...
+status = "generate"
+    [[object.member]]
+    name = "right"
+    version = "1.50"
+    [[object.member]]
+    name = "center"
+    version = "1.50"
+    [[object.member]]
+    name = "decimal"
+    version = "1.50"
```
Comment (sdroege) [inline pango/Gir.toml:328]: You created an issue/MR to mark these enums correctly in C?

File: pango/src/auto/font.rs [hunk 5]
```diff
@@ -27,6 +33,28 @@ glib::wrapper! {
...
 impl Font {
     pub const NONE: Option<&'static Font> = None;
+
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+    #[doc(alias = "pango_font_deserialize")]
+    pub fn deserialize(
+        context: &Context,
+        bytes: &glib::Bytes,
+    ) -> Result<Option<Font>, glib::Error> {
```
Comment (sdroege) [inline pango/src/auto/font.rs:43]: This is not actually `nullable`, is it?

File: pango/src/auto/font.rs [hunk 6]
```diff
@@ -74,6 +102,11 @@ pub trait FontExt: 'static {
     #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_44")))]
     #[doc(alias = "pango_font_has_char")]
     fn has_char(&self, wc: char) -> bool;
+
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+    #[doc(alias = "pango_font_serialize")]
+    fn serialize(&self) -> Option<glib::Bytes>;
```
Comment (sdroege) [inline pango/src/auto/font.rs:109]: And this probably also not `nullable`

File: pango/src/auto/layout.rs [hunk 7]
```diff
@@ -421,6 +427,18 @@ impl Layout {
         }
     }
 
+    #[cfg(any(feature = "v1_50", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+    #[doc(alias = "pango_layout_serialize")]
+    pub fn serialize(&self, flags: LayoutSerializeFlags) -> Option<glib::Bytes> {
```
Comment (sdroege) [inline pango/src/auto/layout.rs:433]: And neither this

File: pango/src/auto/tab_array.rs (hunk 3 above)
Comment (sdroege) [reply pango/src/auto/tab_array.rs:147]: Yes the getters at least shouldn't. Please create an issue :)

File: pango/src/auto/enums.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply pango/src/auto/enums.rs:1014]: my comment was about not having a cfg guard for this but yeah

File: pango/src/auto/layout.rs (hunk 7 above)
Comment (bilelmoussaoui) [reply pango/src/auto/layout.rs:433]: All of those are caused by trust_nullability :/

File: pango/src/auto/tab_array.rs (hunk 3 above)
Comment (bilelmoussaoui) [reply pango/src/auto/tab_array.rs:147]: It seems to break the build, let me fix that :)

File: pango/Gir.toml (hunk 4 above)
Comment (bilelmoussaoui) [reply pango/Gir.toml:327]: Well, as there's no proper annotations for enum members; they are kind of documented already https://docs.gtk.org/Pango/enum.Variant.html

File: pango/src/auto/font.rs (hunk 5 above)
Comment (bilelmoussaoui) [reply pango/src/auto/font.rs:43]: it seems to be nullable https://gitlab.gnome.org/GNOME/pango/-/blob/main/pango/serializer.c#L1543

File: pango/src/auto/layout.rs (hunk 7 above)
Comment (sdroege) [reply pango/src/auto/layout.rs]: Yes, if we opt in to that we need to keep the upstream annotations correct

File: pango/src/enums.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+#[cfg(any(feature = "v1_50", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+use std::fmt;
+
+#[cfg(any(feature = "v1_50", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Clone, Copy)]
+#[non_exhaustive]
+#[doc(alias = "PangoLayoutDeserializeError")]
+pub enum LayoutDeserializeError {
```
Comment (sdroege) [inline pango/src/enums.rs:20]: Why make this manual?

File: pango/Gir.toml (hunk 4 above)
Comment (sdroege) [reply pango/Gir.toml:335]: There is, see https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/230 . We use that extensively in GStreamer.

Comment (bilelmoussaoui) [reply pango/Gir.toml:335]: Oh wow, didn't knew that. So why we always have to set them manually so far? :/

File: pango/src/enums.rs (hunk 8 above)
Comment (bilelmoussaoui) [reply pango/src/enums.rs:20]: See the commit & todo in the code. gir uses a wrong error quark function name (that doesn't even exists??) and we can't disable only that part of the enum

Comment (sdroege) [reply pango/src/enums.rs:20]: Isn't there a configuration for the error quark?

File: pango/Gir.toml (hunk 4 above)
Comment (sdroege) [reply pango/Gir.toml:335]: Because gobject-introspection can only do that since a year :)

File: pango/src/enums.rs (hunk 8 above)
Comment (bilelmoussaoui) [reply pango/src/enums.rs:20]: I looked the config code but didn't find any, ideally the code in library postprocessing needs to be fixed as the quark is properly defined on the gir file.

**Comment by bilelmoussaoui** [comment]
Hmm i messed up the rebase

File: gio/src/auto/flags.rs [hunk 9]
```diff
@@ -2003,6 +2007,8 @@ bitflags! {
         const STDERR_MERGE = ffi::G_SUBPROCESS_FLAGS_STDERR_MERGE as u32;
         #[doc(alias = "G_SUBPROCESS_FLAGS_INHERIT_FDS")]
         const INHERIT_FDS = ffi::G_SUBPROCESS_FLAGS_INHERIT_FDS as u32;
+        #[doc(alias = "G_SUBPROCESS_FLAGS_SEARCH_PATH_FROM_ENVP")]
+        const SEARCH_PATH_FROM_ENVP = ffi::G_SUBPROCESS_FLAGS_SEARCH_PATH_FROM_ENVP as u32;
```
Comment (sdroege) [inline gio/src/auto/flags.rs:2011]: Needs a version flag

File: pango/src/auto/attr_list.rs [hunk 10]
```diff
@@ -86,10 +97,26 @@ impl AttrList {
...
```
Comment (sdroege) [inline pango/src/auto/attr_list.rs:104]: Should probably be a `Result`

File: pango/src/auto/font.rs (hunk 5 above)
Comment (sdroege) [inline pango/src/auto/font.rs:43]: Can this return `None` without an error?

File: pango/src/enums.rs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    MissingValue,
+    #[doc(hidden)]
+    __Unknown(i32),
+}
+
+#[cfg(any(feature = "v1_50", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "v1_50")))]
+impl LayoutDeserializeError {
+    #[doc(alias = "pango_layout_deserialize_error_quark")]
+    pub fn quark() -> Quark {
```
Comment (sdroege) [inline pango/src/enums.rs:35]: Should implement https://gtk-rs.org/gtk-rs-core/git/docs/glib/error/trait.ErrorDomain.html instead

File: pango/src/auto/attr_list.rs (hunk 10 above)
Comment (sdroege) [reply pango/src/auto/attr_list.rs:104]: And implement `FromStr`?

Comment (bilelmoussaoui) [reply pango/src/auto/attr_list.rs:104]: I am pretty sure I did this at some point...meh, got some changes lost during the rebase

Comment (bilelmoussaoui) [reply pango/src/auto/attr_list.rs:104]: Never mind, I did that for TabArray. Will do!

File: pango/src/auto/font.rs (hunk 5 above)
Comment (bilelmoussaoui) [reply pango/src/auto/font.rs:43]: As mentioned in prior comments, it can return None yes

Comment (sdroege) [reply pango/src/auto/font.rs:43]: How very confusing, thanks :)

File: pango/src/enums.rs (hunk 11 above)
Comment (sdroege) [reply pango/src/enums.rs:35]: You do a few lines below, but having a separate function for the quark is useless

File: pango/Gir.toml (hunk 4 above)
Comment (bilelmoussaoui) [reply pango/Gir.toml:339]: https://gitlab.gnome.org/GNOME/pango/-/merge_requests/575

File: pango/Gir.toml [hunk 12]
```diff
@@ -310,3 +324,48 @@ boxed_inline = true
...
+
+[[object]]
+name = "Pango.TabArray"
+status = "generate"
+    [[object.function]]
+    pattern = "(to_string|get_decimal_point|get_positions_in_pixels|get_tab)"
+        [[object.function.parameter]]
+        pattern = "*" # because we can't target self
```
Comment (sdroege) [inline pango/Gir.toml:347]: Can't we? It probably works if you use the parameter name from the C function instead of `self`

**Review by sdroege — APPROVED** [review]

**Pull request merged**